tests/test_utils.py

```diff
@@ -0,0 +1,59 @@
+import dask
+import pytest
+from distributed import Client
+from prefect import flow, task
+
+from prefect_dask import DaskTaskRunner, get_dask_client
+
+
+@pytest.mark.parametrize("timeout", [None, 10])
+def test_get_dask_client_task_run_context_integration(timeout):
+    @task
+    def test_task():
+        delayed_num = dask.delayed(42)
+        with get_dask_client(timeout=timeout) as client:
+            assert isinstance(client, Client)
+            if timeout is not None:
+                assert client._timeout == timeout
+            result = client.compute(delayed_num).result()
+        return result
+
+    @flow(task_runner=DaskTaskRunner)
+    def test_flow():
+        future = test_task.submit()
+        return future.result()
+
+    assert test_flow() == 42
+
+
+def test_get_dask_client_flow_run_context_integration():
+    @flow(task_runner=DaskTaskRunner)
+    def test_flow():
+        delayed_num = dask.delayed(42)
+        with get_dask_client() as client:
+            assert isinstance(client, Client)
+            result = client.compute(delayed_num, sync=True)
```

Comment:
Are we going to make users pass `sync=True` to all calls? Can we just get a new client depending on if the flow is async/sync?

Reply:
I think it'd be confusing if we start a new client because the dashboard link is not the same.

Reply:
Isn't the dashboard link determined by the server? We'd need to get a new client for the existing server.

Reply:
We can note this as a future optimization if you prefer, seems hard.

Reply:
Not sure I follow, but here's the output:
```
from dask.distributed import Client
client = Client()
new_client = Client()
print(client.dashboard_link, new_client.dashboard_link)
```
Outputs:
`http://127.0.0.1:60725/status http://127.0.0.1:8787/status`

Reply:
Perhaps we can make `self._connect_to`

Reply:
Is there a way to check if the flow run is async or not? I checked flow run context and also tried using `in_async_main_thread() or in_async_worker_thread()` which still returns True if the flow is not async.

Reply:
Maybe `asynchronous = sync_portal is not None`

Reply:
The issue is that the code you shared is creating a second temporary Dask cluster. You'd need to connect to the existing one. I presume there's some way in Dask to connect to the same cluster that an existing client has created.

> Is there a way to check if the flow run is async or not?

`FlowRunContext.flow.isasync` should do it

Reply:
Hey folks, just chiming in here to say that @madkinsz is right. You can make any _N_ `Client` instances pointing to the same cluster and they'll all have the same dashboard address.

for example:

```python
from distributed import Client
from coiled import Cluster

coiled_cluster = Cluster(n_workers=4)
client_1 = Client(coiled_cluster)
client_2 = Client(coiled_cluster)

print(f"{client_1.dashboard_link=}")
print(f"{client_2.dashboard_link=}")
```

prints:

```text
client_1.dashboard_link='http://3.90.45.123:8787/'
client_2.dashboard_link='http://3.90.45.123:8787/'
```

Reply:
Oh and of course:

```python
client_3 = Client(client_1.cluster)
print(f"{client_3.dashboard_link=}")
```

```
client_3.dashboard_link='http://3.90.45.123:8787/'
```

Reply:
Right, so I tried to create [`self._connect_to`](https://github.com/PrefectHQ/prefect-dask/pull/33/files#diff-e48979d13e982d1ac163b900fac9a9d8cd8366b352dbc4b89b17d5b21de4120eR280) but I don't think I can [pickle the Cluster object](https://github.com/PrefectHQ/prefect-dask/actions/runs/3146599021/jobs/5115226101#step:5:486)

When I try to `get_client` I get:
`ValueError: No global client found and no address provided`

Reply:
Thanks phobson!

> Right, so I tried to create [self._connect_to](https://github.com/PrefectHQ/prefect-dask/pull/33/files#diff-e48979d13e982d1ac163b900fac9a9d8cd8366b352dbc4b89b17d5b21de4120eR280) but I don't think I can [pickle the Cluster object](https://github.com/PrefectHQ/prefect-dask/actions/runs/3146599021/jobs/5115226101#step:5:486)

You can include the cluster on the task runner then just drop it when it is pickled. We only need it in the flow context and there's no serialization happening there.

> Unless there's a way to fix sync_compatible in this case

It's a trivial fix for `sync_compatible`, we just need to resolve the referenced function with `__wrapped__` (or whatever the decorator attribute is) until we get the actual function when checking if an async function was used.

Reply:
Not sure if it's trivial:
```
def sync_compatible(async_fn: T) -> T:
    base_async_fn = async_fn
    while hasattr(base_async_fn, "__wrapped__"):
        base_async_fn = base_async_fn.__wrapped__
        print(base_async_fn.__name__)

    if not inspect.iscoroutinefunction(base_async_fn):
        raise TypeError("The decorated function must be async.")
```


```
from asyncio import sleep
from contextlib import asynccontextmanager


@asynccontextmanager
@sync_compatible
async def test_something():
    await sleep(1)
    print('hey')

test_something()
```

Results in:
```


RuntimeError: A 'sync_compatible' method was called from a context that was previously async but is now sync. The sync call must be changed to run in a worker thread to support sending the coroutine for 'test_something' to the main thread.
```

Reply:
Trivial to fix detection I thought :D turns out `iscoroutinefunction` says async generators aren't coroutine functions! Also turns out it's not so trivial to support an async context manager from a sync context. I'm a little too busy to try to sort it out right now. Here's some `sync_compatible` improvements https://github.com/PrefectHQ/prefect/pull/7009 but we should just define two utilities, I think.